syntax = "proto3";

package telos;

option go_package = "github.com/nevinshine/telos-runtime/shared/pb";

// The "Cortex" acts as the central server.
service TelosControl {
    // 1. Browser Eye reports finding hidden text/taint
    rpc ReportTaint (TaintReport) returns (Ack);

    // 2. Agent declares what it wants to do (Intent)
    rpc DeclareIntent (IntentRequest) returns (IntentVerdict);

    // 3. Daemons (Core/Edge) ask for latest rules
    rpc GetPolicy (PolicyQuery) returns (PolicyRules);
}

// --- MESSAGES ---

message Ack {
    bool success = 1;
    string message = 2;
}

message TaintReport {
    string source_id = 1;      // e.g., "tab_ID_123"
    int32 pid = 2;             // Process ID of the browser
    string url = 3;            // URL where taint was found
    TaintLevel level = 4;
    string payload_preview = 5; // First 64 chars of the hidden text
}

message IntentRequest {
    int32 agent_pid = 1;
    string natural_language_goal = 2; // "I want to download invoices"
    repeated string planned_actions = 3; // ["connect:billing.com:443"]
}

message IntentVerdict {
    bool allowed = 1;
    string reason = 2;
    int64 policy_ttl_ms = 3;   // How long the firewall stays open
}

message PolicyQuery {
    int32 pid = 1;
}

message PolicyRules {
    TaintLevel max_allowed_taint = 1;
    repeated string allowed_ips = 2;
    repeated string allowed_paths = 3; // e.g., "/tmp/downloads/*"
}

enum TaintLevel {
    CLEAN = 0;
    LOW = 1;       // Unverified Web
    MEDIUM = 2;    // Hidden DOM Elements
    HIGH = 3;      // Known Injection Signature
    CRITICAL = 4;  // Shellcode / Payload Detected
}
